lib/srv/db/mysql/gcp.go

```diff
@@ -0,0 +1,138 @@
+/*
+ * Teleport
+ * Copyright (C) 2024  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package mysql
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/lib/cloud/gcp"
+	"github.com/gravitational/teleport/lib/defaults"
+	"github.com/gravitational/teleport/lib/services"
+	"github.com/gravitational/teleport/lib/srv/db/common"
+)
+
+func isDBUserGCPServiceAccount(dbUser string) bool {
+	if strings.Contains(dbUser, "@") {
+		switch {
+		// Example: mysql-iam-user@my-project-id.iam
+		// This format is used to align with PostgreSQL.
+		case strings.HasSuffix(dbUser, ".iam"):
+			return true
+		// Example: mysql-iam-user@my-project-id..gserviceaccount.com
+		case strings.HasSuffix(dbUser, ".iam.gserviceaccount.com"):
+			return true
+		}
+	}
+	return false
+}
+
+func gcpServiceAccountToDatabaseUser(serviceAccountName string) string {
+	user, _, _ := strings.Cut(serviceAccountName, "@")
+	return user
+}
+
+func databaseUserToGCPServiceAccount(sessionCtx *common.Session) string {
+	return fmt.Sprintf("%s@%s.iam.gserviceaccount.com", sessionCtx.DatabaseUser, sessionCtx.Database.GetGCP().ProjectID)
+}
+
+func (e *Engine) getGCPUserAndPassword(ctx context.Context, sessionCtx *common.Session, gcpClient gcp.SQLAdminClient) (string, string, error) {
+	// If `--db-user` is an service account email, use IAM Auth.
+	if isDBUserGCPServiceAccount(sessionCtx.DatabaseUser) {
+		return e.getGCPIAMUserAndPassword(ctx, sessionCtx)
+	}
+
+	// Get user info to decide how to authenticate.
+	user, err := gcpClient.GetUser(ctx, sessionCtx.Database, sessionCtx.DatabaseUser)
```

Comment:
This fallback will introduce a divergence between the PostgreSQL IAM authentication flow and the MySQL authentication flow."

As far as I know, PostgreSQL IAM authentication requires the --db-user=user@my-project-id.iam format. Could you leave a todo to align the comment with the PostgreSQL implementation?

Reply:
This one is tricky.

For GCP Postgres, I do recall from support that some customers asked to connect without the `@<my-project-id>.iam` part.  However, for GCP Postgres IAM auth user, `<my-name>@<my-project-id>.iam` is the actual in-database username, which uses the service account `<my-name>@<my-project-id>.iam.gserviceaccount.com`. So with `@<my-project-id>.iam` is technically more correct. 

For GCP MySQL, the in-database username is `<my-name>` when using the service account `<my-name>@<my-project-id>.iam.gserviceaccount.com`. I added `<my-name>@<my-project-id>.iam` and  `<my-name>@<my-project-id>.iam.gserviceaccount.com` for GCP MySQL for convenience (in case there is no `GetUser` permission). But now I am not so sure if that makes sense.

To summerize:
| Type | In-Database user | Service account |
| ----- | ----------------- | ---------------- |
| Postgres  | `<my-name>@<my-project-id>.iam`  | `<my-name>@<my-project-id>.iam.gserviceaccount.com` |
| MySQL  | `<my-name>`  | `<my-name>@<my-project-id>.iam.gserviceaccount.com` |

I think we have several approaches:
1. First is to restrict `--db-user` to match in-database username. So `<my-name>@<my-project-id>.iam` for Postgres, and just `<my-name>` for MySQL
2. Keep Postgres as it is, for MySQL, accept either in-database username or the full name. Do NOT accept Postgres's format.
3. For best convenience, accept all three formats for both databases. 

I am more leaning towards option 2 now. What do you think?

Reply:
> I am more leaning towards option 2 now. What do you think?

Seems reasonable 👍